src/core/raster/qgsrasterinterface.cpp

```diff
@@ -363,8 +363,8 @@ void QgsRasterInterface::initHistogram( QgsRasterHistogram &histogram,
       }
       else
       {
-        // This is for not integer types:
-        myBinCount = std::min( 2000, histogram.width * histogram.height );
+        // This is for not integer types where we cannot limit the bin size
+        myBinCount = histogram.width * histogram.height;
```

Comment:
not related to this change, but I see we could potentially have a integer overflow on multiplication on sufficiently large raster when sampleSize == 0 and ( capabilities() & Size ). Same just above at line 362. At line 309 too

Reply:
`myHistogram.histogramVector.resize( myBinCount );` at line 430 could throw a out-of-memory exception if myBinCount is too large. Perhaps 2000 is too small, but unbounded bincount could be a memory & performance killer. Maybe 1 million or 10 million upper limit ?

Reply:
> not related to this change, but I see we could potentially have a integer overflow on multiplication on sufficiently large raster when sampleSize == 0 and ( capabilities() & Size ). Same just above at line 362. At line 309 too

Are you sure? https://github.com/qgis/QGIS/pull/37457/files/d49e140cf66468f9743fe2cb05b4f155898e37f8#diff-1a8a563373bea4e36bc554b9ff1c0067R309 these are all already `double`.

Ok for line 362.

Please see my last commit for the limits.

Reply:
> Are you sure? https://github.com/qgis/QGIS/pull/37457/files/d49e140cf66468f9743fe2cb05b4f155898e37f8#diff-1a8a563373bea4e36bc554b9ff1c0067R309 these are all already `double`

at line 309, ( finalExtent.width() * finalExtent.height() ) / sampleSize is evaluated as a integer. std::sqrt() transforms to double only after. Casting finalExtent.width() to double should avoid any int overflow